### PR TITLE
[SmartSwitch] Define DPU helper functions

### PR DESCRIPTION
**Pull request by vvolam**
#### Why I did it
Add new DPU helper functions to avoid code duplication

##### Work item tracking
- Microsoft ADO **(number only)**: 30158318 

#### How I did it
- Define new is_dpu, get_num_dpus helpers.
- Define get_platform_json_data and get_dpu_info helper function for common code and refactor code accordingly.

#### How to verify it

Verified the changes on smart switch

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

File: src/sonic-py-common/sonic_py_common/device_info.py [hunk 1]
```diff
@@ -582,15 +610,30 @@ def is_smartswitch():
...
+    if not platform:
+        return False
+
+    if not is_smartswitch():
         return False
 
+    # Retrieve platform.json data
+    platform_data = get_platform_json_data()
+    if platform_data:
+        return "DPU" in platform_data
```
Comment (prgeor) [inline src/sonic-py-common/sonic_py_common/device_info.py:633]: @vvolam platform.json file on the DPU will have DPU? Is this aligned to smartswitch HLD?

File: src/sonic-py-common/sonic_py_common/device_info.py [hunk 2]
```diff
@@ -582,15 +610,30 @@ def is_smartswitch():
...
-    except IOError:
+    # Retrieve platform.json data
+    platform_data = get_platform_json_data()
+    if platform_data:
+        return "DPUS" in platform_data
+
+    return False
+
+
+def is_dpu():
```
Comment (prgeor) [inline src/sonic-py-common/sonic_py_common/device_info.py:621]: @vvolam can you have something like here? https://github.com/sonic-net/sonic-utilities/blob/master/counterpoll/main.py

File: src/sonic-py-common/sonic_py_common/device_info.py (hunk 1 above)
Comment (oleksandrivantsiv) [reply src/sonic-py-common/sonic_py_common/device_info.py:633]: @prgeor it is based on https://github.com/sonic-net/SONiC/blob/master/doc/smart-switch/ip-address-assigment/smart-switch-ip-address-assignment.md#dpu-platformjson

File: src/sonic-py-common/sonic_py_common/device_info.py (hunk 2 above)
Comment (oleksandrivantsiv) [reply src/sonic-py-common/sonic_py_common/device_info.py:621]: `switch_type` is take from the config DB. It might not be available at the early boot stage. It is better to rely on `DPU` attribute from the `platform.json` since it is static configuration that is always available

File: src/sonic-py-common/sonic_py_common/device_info.py [hunk 3]
```diff
@@ -892,37 +935,54 @@ def is_frontend_port_present_in_host():
...
+        return {}
 
-    # Get Platform path.
-    platform_path = get_path_to_platform_dir()
+    # Retrieve platform.json data
+    platform_data = get_platform_json_data()
+    if platform_data:
+        # Convert keys to lower case to avoid case sensitivity.
+        data = {k.lower(): v for k, v in platform_data.items()}
+        dpu_info = data.get('dpus', {})
```
Comment (oleksandrivantsiv) [inline src/sonic-py-common/sonic_py_common/device_info.py:955]: Can we make this function compatible with the DPU? E.g.:
```
if 'DPUS' in platform_data:
    return platform_data['DPUS']
elif 'DPU' in platform_data:
    return platform_data['DPU']
```

File: src/sonic-py-common/sonic_py_common/device_info.py [hunk 4]
```diff
@@ -892,37 +935,54 @@ def is_frontend_port_present_in_host():
...
+    dpu_info = get_dpu_info()
+    if dpu_info is not None and len(dpu_info) > 0:
+        return len(dpu_info)
 
     return 0
+
+
+def get_dpu_list():
+    """
+    Retrieves the list of DPUs from platform.json file.
```
Comment (oleksandrivantsiv) [inline src/sonic-py-common/sonic_py_common/device_info.py:978]: Why is this needed? What is the difference from `get_dpu_info`?

File: src/sonic-py-common/sonic_py_common/device_info.py (hunk 1 above)
Comment (vvolam) [reply src/sonic-py-common/sonic_py_common/device_info.py:633]: Thank you @oleksandrivantsiv

File: src/sonic-py-common/sonic_py_common/device_info.py (hunk 4 above)
Comment (vvolam) [reply src/sonic-py-common/sonic_py_common/device_info.py:978]: get_dpu_list() only returns the list of DPU module names on NPU, whereas get_dpu_info() returns both the DPU module information and the values for all the DPU modules.

File: src/sonic-py-common/sonic_py_common/device_info.py (hunk 3 above)
Comment (vvolam) [reply src/sonic-py-common/sonic_py_common/device_info.py:955]: Yes, good! I will modify it.

File: src/sonic-py-common/sonic_py_common/device_info.py (hunk 4 above)
Comment (vvolam) [reply src/sonic-py-common/sonic_py_common/device_info.py:978]: For example, if platform.json is 
"{
    "DPUS": {
        "dpu0": {
            "midplane_interface": "dpu0"
        },
        "dpu1": {
            "midplane_interface": "dpu1"
        }
    }
}"

get_dpu_info() returns
{
    "dpu0": {
        "midplane_interface": "dpu0"
    },
    "dpu1": {
        "midplane_interface": "dpu1"
    }
}

and get_dpu_list() return this
["dpu0", "dpu1"]

Comment (prgeor) [reply src/sonic-py-common/sonic_py_common/device_info.py:981]: @vvolam can you specify the return accordingly in the code comment?

File: src/sonic-py-common/sonic_py_common/device_info.py (hunk 3 above)
Comment (prgeor) [reply src/sonic-py-common/sonic_py_common/device_info.py]: @vvolam what is the intention of this API? Previously, you cared about DPUs which indicated this was supposed to be used by NPU ...but now it usage is on DPU as well?

Comment (vvolam) [reply src/sonic-py-common/sonic_py_common/device_info.py]: @prgeor Intention of this API is to return DPU information from the platform.json file. In case of NPU, it returns the information for all DPUs whereas in case of DPU, it returns only the DPU specific information.

File: src/sonic-py-common/sonic_py_common/device_info.py (hunk 4 above)
Comment (vvolam) [reply src/sonic-py-common/sonic_py_common/device_info.py:981]: Added an example. Thank you!

**Review by prgeor — DISMISSED** [review]

**Comment by vvolam** [comment]
> @vvolam please check the build errors

@prgeor There is a known build issue which got fixed through https://github.com/sonic-net/sonic-buildimage/pull/20791. I have restarted the pipeline after the fix.

**Comment by vvolam** [comment]
> no unit test?

@lguohan

As coverage check is not present for sonic-buildimage, lot of PRs are missing unit tests which confused me. I have added unittests for all smartswitch function in device_info.py file. Thank you for pointing this.

I will work to add coverage check for sonic-buildimage PRs as well.

**Review by prgeor — APPROVED** [review]

**Review by yxieca — APPROVED** [review]

File: src/sonic-py-common/sonic_py_common/device_info.py [hunk 5]
```diff
@@ -892,37 +935,61 @@ def is_frontend_port_present_in_host():
...
+    # Retrieve platform.json data
+    platform_data = get_platform_json_data()
+    if not platform_data:
+        return {}
+
+    if "DPUS" in platform_data:
+        return platform_data["DPUS"]
+    elif 'DPU' in platform_data:
+        return platform_data['DPU']
+
```
Comment (prgeor) [inline src/sonic-py-common/sonic_py_common/device_info.py:959]: @vvolam `else` case missing?

Comment (vvolam) [reply src/sonic-py-common/sonic_py_common/device_info.py:961]: @prgeor My bad, added it. Thank you!

**Review by prgeor — APPROVED** [review]

**Review by Pterosaur — APPROVED** [review]
